### PR TITLE
feat: add and port to SnapshotManager

### PR DESCRIPTION
**Pull request by mogery**
Supersedes #3430 (part 1)

This PR centralises all snapshot file management code in a SnapshotManager struct, and ports everything to use it. This is needed to provide an abstraction for snapshot files, so that we can store them not only locally, but also on S3 (will be introduced in a future PR).

It also introduces a SnapshotFile struct which replaces `PathBuf`s that were used to refer to snapshots before.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by mogery** [comment]
@generall This is -- in my opinion -- the smallest PR that makes sense and doesn't break build. It might still be too big, but it doesn't include the S3 code, so that's one less thing to test and review.

**Review by generall — CHANGES_REQUESTED** [review]
Hey @mogery, thanks for PR! I think overall it is moving into the right directlion, however I think it might benefit from making 
 `SnapshotManager` less global and more independent entity. Here is how I can see that:

- `SnapshotManager` doesn't really have any internal state. It can be easily created from the configuration (even with s3 support enabled).
All relevant structures in the code (ToC, Collection, Shard) have access to storage configuration, so it should be easy to build `SnapshotManager` on demang.

- Current `SnapshotManager` knows too much about the whole project strucutre. E.g. it knows about the concept of shard_ids and collections. At the same time Collection should know about `SnapshotManager` which either create circular dependency or forces us to drop proper types at compile time (see the commend about u32 as shard is). I propose to make `SnapshotManager` more local, let it only work with single given "directory" and abstract it away from differenciation of the snapshot types.

Those two chanes should allow making PR much more compact, avoid extra pagameters in seemengly unrelated function, and also should allow to test Snapshot manager independently from the ToC and other storage structures.

**Comment by mogery** [comment]
Thank you for the review @generall!

> I think overall it is moving into the right direction, however I think it might benefit from making  `SnapshotManager` less global and more independent entity. Here is how I can see that:
> 
> - `SnapshotManager` doesn't really have any internal state. It can be easily created from the configuration (even with s3 support enabled).
> 
> All relevant structures in the code (ToC, Collection, Shard) have access to storage configuration, so it should be easy to build `SnapshotManager` on demand.
> 

Great catch! I didn't realize that but I totally agree. 

> 
> - Current `SnapshotManager` knows too much about the whole project structure. E.g. it knows about the concept of shard_ids and collections. At the same time Collection should know about `SnapshotManager` which either create circular dependency or forces us to drop proper types at compile time (see the commend about u32 as shard is). I propose to make `SnapshotManager` more local, let it only work with single given "directory" and abstract it away from differenciation of the snapshot types.


I think we can do it even better: in the S3 code to handle `SnapshotFile::OutOfPlace` correctly I added some code that determines if the path should point to S3 or not. All it did is check if the provided path is inside the configured snapshot path (or a subdirectory). I think we can extend this to all operations, remove `SnapshotFile` completely, and rely on paths entirely again, which would simplify things and reduce `SnapshotManager`'s dependency on the current structure. Then the S3-specific logic would only kick in if S3 is enabled and the path should point to S3.

File: lib/segment/Cargo.toml [hunk 1]
```diff
@@ -62,7 +62,7 @@ smol_str = "0.2.1"
 lazy_static = "1.4"
 
 sysinfo = "0.30"
-charabia = { version = "0.8.5", default-features = false, features = ["greek", "hebrew", "thai"] }
```
[the remark this answers is not in the document]
Comment (mogery) [reply lib/segment/Cargo.toml:65]: Oh absolutely, that's an accidental diff I left in. For some reason build was failing for me without that feature.

File: lib/common/snapshot_manager/src/file.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
+use std::path::PathBuf;
+
+#[derive(Debug, Clone)]
+pub enum SnapshotFile {
+    Normal {
+        name: String,
+        collection: Option<String>,
+        shard: Option<u32>,
```
[the remark this answers is not in the document]
Comment (mogery) [reply lib/common/snapshot_manager/src/file.rs:8]: I understand what you're saying, but I still think that `SnapshotFile` is a super useful helper that concentrates all the path-building logic and structure into one consistent and easily editable place (if the structure needs to change).

**Comment by mogery** [comment]
@generall Made the following changes:
- `SnapshotFile` has been deleted.
- `SnapshotManager` has been de-globalized, instead it is instantiated on demand in many different places.
- `SnapshotManager` is now scoped to a single directory with `SnapshotManager::scope`. It is unaware of the project structure, other code scopes it to certain directories to achieve certain functionality. (for example, to list all collection snapshots: `toc.snapshot_manager().scope("my_collection/").do_list_snapshots().await?`)
- `SnapshotManager` functions that refer to exact snapshots (e.g. `do_delete_snapshot`) take an `impl AsRef<Path>` as the "snapshot name". This allows for the normal behaviour if a relative path like `test_snapshot.snapshot` is given, but also `OutOfPlace` behaviour is supported if an absolute path like `/path/to/my/snapshot.snapshot` is given. This allows for versatility while still centralizing snapshot access code. This can also be easily adapted to support S3.

I didn't have time to run any integration tests, so I'd appreciate a CI run. :)

File: lib/segment/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mogery) [reply lib/segment/Cargo.toml]: Removed.

File: lib/common/snapshot_manager/src/file.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mogery) [reply lib/common/snapshot_manager/src/file.rs]: Resolved.

File: src/main.rs [hunk 3]
```diff
@@ -164,6 +167,7 @@ fn main() -> anyhow::Result<()> {
             temp_path,
             &full_snapshot,
             &settings.storage.storage_path,
+            snapshot_manager.clone(),
```
[the remark this answers is not in the document]
Comment (mogery) [reply src/main.rs]: Resolved in abc8b3ad7eb6b3689afdc8dd0f154b7625a478da

**Comment by mogery** [comment]
Fixed the failing tests.

File: lib/collection/src/shards/transfer/snapshot.rs [hunk 4]
```diff
@@ -158,7 +159,7 @@ pub(super) async fn transfer_snapshot(
     remote_shard: RemoteShard,
     channel_service: ChannelService,
     consensus: &dyn ShardTransferConsensus,
-    snapshots_path: &Path,
+    snapshot_manager: SnapshotManager,
```
Comment (generall) [inline lib/collection/src/shards/transfer/snapshot.rs:162]: Why `snapshot_manager` is required here as an argument, while we have access to `shard_holder` and `shard_id`?

**Comment by generall** [comment]
Hey @mogery, unfortunately the PR still have too big impact on the internals, which makes it very hard too review and validate.

There are many places where `snapshot_manager` is introduces as a new function argument, even though it doesn't provide any additional context to the function. 

I propose to remove all of those places and try to keep PR at least within 1000 lines of code (ideally within 500 lines)

File: lib/collection/src/shards/transfer/snapshot.rs (hunk 4 above)
Comment (mogery) [reply lib/collection/src/shards/transfer/snapshot.rs:162]: AFAICT, `shard_holder` doesn't have a path to access any pre-existing `SnapshotManager` instance, or a config struct. It cannot instantiate a `SnapshotManager` on its own.

Comment (generall) [reply lib/collection/src/shards/transfer/snapshot.rs:162]: `shard_holder` have shards, and shards have config

**Comment by generall** [comment]
Closing in favor of https://github.com/qdrant/qdrant/pull/4150

**Pull request closed without merging**